File: backend/app/render/photo_layout.py
"""
How the survey photographs are laid out in the report.

The client builds his photo pages with his own Auto-on-render tool (Pro mode),
and the layout here is that tool's, measured against his reports: in 60 of his
perishable reports, 50 carry 8 photos to an A4 page and 10 carry 6; the photos
are 8.2 x 5.6 cm (the tool's 310 x 210 px), two across with no gap, no border,
and the caption under each is just "Survey Photo No. N".

The same options the tool offers — border and its colour, caption wording,
numbering start, caption font, size and colour, photo compression — are kept on
the photo block as `layout`, and both the Word and the HTML renderer read them
from here so the two stay alike.
"""

from __future__ import annotations

import re
from typing import Any, Dict, List

# The tool's Pro-mode photo box, in pixels at 96 dpi: 8.20 x 5.56 cm.
IMAGE_W_PX = 310
IMAGE_H_PX = 210
EMU_PER_PX = 9525
CM_PER_PX = 2.54 / 96

# Twips, as the tool sets them.
CELL_MARGIN = {"plain": {"top": 5, "bottom": 5, "left": 20, "right": 20},
               "border": {"top": 10, "bottom": 10, "left": 10, "right": 10}}
PARA_SPACING = {"plain": 5, "border": 10}
CAPTION_MARGIN = {"top": 5, "bottom": 5, "left": 20, "right": 20}
BORDER_SIZE = 12  # eighths of a point

KEYWORDS = ("Survey Photo No.", "QC Inspection Photo No.")
FONTS = ("Arial", "Calibri", "Times New Roman", "Georgia", "Verdana")
FONT_SIZES = (8, 9, 10, 11, 12, 14, 16, 18, 20)
FONT_COLORS = ("000000", "FFFFFF", "000080", "404040", "006400")
BORDER_COLORS = ("000000", "808080", "000080", "8B4513")
PER_PAGE = (8, 6)

# Longest side in px and JPEG quality, as the tool's presets. "original"
# keeps the full size; only the phone's rotation is corrected.
QUALITY = {
    "original": (None, 95),
    "high": (3000, 88),
    "balanced": (2200, 78),
    "small": (1600, 65),
    "display": (800, 80),   # HTML preview only
}

DEFAULTS: Dict[str, Any] = {
    "per_page": 8,
    "border": False,
    "border_color": "000000",
    "caption_keyword": "Survey Photo No.",
    "number_from": 1,
    "caption_font": "Arial",
    "caption_size": 11,
    "caption_color": "000000",
    "quality": "balanced",
    "landscape": True,
    "show_heading": False,
}


def layout_of(block: Dict[str, Any]) -> Dict[str, Any]:
    """The block's layout with anything missing or out of range set to the default."""
    raw = block.get("layout") or {}
    out = dict(DEFAULTS)

    def pick(key, allowed, cast=lambda v: v):
        try:
            v = cast(raw.get(key))
        except (TypeError, ValueError):
            return
        if v in allowed:
            out[key] = v

    pick("per_page", PER_PAGE, int)
    pick("caption_font", FONTS)
    pick("caption_size", FONT_SIZES, int)
    pick("caption_color", FONT_COLORS, lambda v: str(v).upper().lstrip("#"))
    pick("border_color", BORDER_COLORS, lambda v: str(v).upper().lstrip("#"))
    pick("quality", tuple(k for k in QUALITY if k != "display"))
    for key in ("border", "landscape", "show_heading"):
        if isinstance(raw.get(key), bool):
            out[key] = raw[key]
    kw = " ".join(str(raw.get("caption_keyword") or "").split())
    if kw:
        out["caption_keyword"] = kw[:40]
    try:
        n = int(raw.get("number_from"))
        if 1 <= n <= 9999:
            out["number_from"] = n
    except (TypeError, ValueError):
        pass
    return out


_PREFIX_ONLY = re.compile(
    r"^\(?\s*(?:survey\s+|qc\s+inspection\s+)?(?:photo|image|pic|img)\s*(?:no\.?)?\s*\d+\s*\)?[:\-—\.,\s]*$",
    re.I,
)
_PREFIX = re.compile(
    r"^\s*\(?\s*(?:survey\s+|qc\s+inspection\s+)?(?:photo|image|pic|img)\s*(?:no\.?)?\s*\d+\s*\)?\s*[:\-—\.,\s]\s*",
    re.I,
)


def strip_number(text: str) -> str:
    """
    The caption text without a photo number typed into it.

    Numbers are worked out when the report is made. Older reports and the
    photo studio wrote them into the text ("Photo 1", "Survey Photo No. 3: Seal"),
    which printed twice once the number was added again.
    """
    t = (text or "").strip()
    if not t or _PREFIX_ONLY.match(t):
        return ""
    return _PREFIX.sub("", t, count=1).strip()


def caption(layout: Dict[str, Any], number: int, text: str = "") -> str:
    body = strip_number(text)
    head = f"{layout['caption_keyword']} {number}"
    return f"{head} — {body}" if body else head


def photos_of(block: Dict[str, Any], computed: Dict[str, Any], assets: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Every photo in the block, in order, with its printed number and caption."""
    lay = layout_of(block)
    shift = lay["number_from"] - 1
    groups_c = (computed or {}).get("groups", {})
    out: List[Dict[str, Any]] = []
    for g in block.get("groups", []) or []:
        ids = g.get("asset_ids", []) or []
        numbers = groups_c.get(g.get("id", ""), {}).get("numbers") or list(range(len(out) + 1, len(out) + 1 + len(ids)))
        for aid, num in zip(ids, numbers):
            n = num + shift
            out.append({
                "asset_id": aid,
                "number": n,
                "caption": caption(lay, n, g.get("observation", "")),
                "asset": assets.get(aid, {}) or {},
            })
    return out


def pages(photos: List[Any], layout: Dict[str, Any]) -> List[List[Any]]:
    """Photos split into pages; a heading on the first page takes one row."""
    per = layout["per_page"]
    first = per - 2 if layout["show_heading"] else per
    if not photos:
        return []
    chunks = [photos[:first]]
    rest = photos[first:]
    chunks += [rest[i:i + per] for i in range(0, len(rest), per)]
    return [c for c in chunks if c]

File: frontend/src/utils/photoLayout.ts
/**
 * How the survey photographs are laid out in the report — the same rules as
 * backend/app/render/photo_layout.py, so the preview shows what Word prints.
 *
 * The layout is the client's own photo tool (Auto-on-render, Pro mode), as
 * measured in his reports: A4, two across, 8.2 x 5.6 cm each, no border,
 * "Survey Photo No. N" under each, 8 to a page (sometimes 6).
 */

export interface PhotoLayout {
  per_page: 8 | 6;
  border: boolean;
  border_color: string;
  caption_keyword: string;
  number_from: number;
  caption_font: string;
  caption_size: number;
  caption_color: string;
  quality: 'original' | 'high' | 'balanced' | 'small';
  landscape: boolean;
  show_heading: boolean;
}

export const PHOTO_DEFAULTS: PhotoLayout = {
  per_page: 8,
  border: false,
  border_color: '000000',
  caption_keyword: 'Survey Photo No.',
  number_from: 1,
  caption_font: 'Arial',
  caption_size: 11,
  caption_color: '000000',
  quality: 'balanced',
  landscape: true,
  show_heading: false,
};

export const KEYWORDS = ['Survey Photo No.', 'QC Inspection Photo No.'];
export const FONTS = ['Arial', 'Calibri', 'Times New Roman', 'Georgia', 'Verdana'];
export const FONT_SIZES = [8, 9, 10, 11, 12, 14, 16, 18, 20];
export const FONT_COLORS = [
  { name: 'Black', value: '000000' },
  { name: 'White', value: 'FFFFFF' },
  { name: 'Navy Blue', value: '000080' },
  { name: 'Dark Gray', value: '404040' },
  { name: 'Dark Green', value: '006400' },
];
export const BORDER_COLORS = [
  { name: 'Black', value: '000000' },
  { name: 'Gray', value: '808080' },
  { name: 'Navy', value: '000080' },
  { name: 'Brown', value: '8B4513' },
];
export const QUALITIES: { key: PhotoLayout['quality']; label: string; note: string }[] = [
  { key: 'original', label: 'Original', note: 'Full size — large Word file' },
  { key: 'high', label: 'High', note: '~2 MB a photo' },
  { key: 'balanced', label: 'Balanced', note: '~1 MB a photo' },
  { key: 'small', label: 'Small', note: '~500 KB a photo' },
];

/** The photo box in the report: 310 x 210 px at 96 dpi. */
export const PHOTO_W_CM = (310 * 2.54) / 96;
export const PHOTO_H_CM = (210 * 2.54) / 96;

export function layoutOf(block: any): PhotoLayout {
  const raw = block?.layout || {};
  const out: PhotoLayout = { ...PHOTO_DEFAULTS };
  if (Number(raw.per_page) === 6 || Number(raw.per_page) === 8) out.per_page = Number(raw.per_page) as 8 | 6;
  if (FONTS.includes(raw.caption_font)) out.caption_font = raw.caption_font;
  if (FONT_SIZES.includes(Number(raw.caption_size))) out.caption_size = Number(raw.caption_size);
  const hex = (v: any) => String(v || '').toUpperCase().replace(/^#/, '');
  if (FONT_COLORS.some((c) => c.value === hex(raw.caption_color))) out.caption_color = hex(raw.caption_color);
  if (BORDER_COLORS.some((c) => c.value === hex(raw.border_color))) out.border_color = hex(raw.border_color);
  if (QUALITIES.some((q) => q.key === raw.quality)) out.quality = raw.quality;
  (['border', 'landscape', 'show_heading'] as const).forEach((k) => {
    if (typeof raw[k] === 'boolean') out[k] = raw[k];
  });
  const kw = String(raw.caption_keyword || '').split(/\s+/).filter(Boolean).join(' ');
  if (kw) out.caption_keyword = kw.slice(0, 40);
  const n = parseInt(raw.number_from, 10);
  if (n >= 1 && n <= 9999) out.number_from = n;
  return out;
}

const PREFIX_ONLY = /^\(?\s*(?:survey\s+|qc\s+inspection\s+)?(?:photo|image|pic|img)\s*(?:no\.?)?\s*\d+\s*\)?[:\-—.,\s]*$/i;
const PREFIX = /^\s*\(?\s*(?:survey\s+|qc\s+inspection\s+)?(?:photo|image|pic|img)\s*(?:no\.?)?\s*\d+\s*\)?\s*[:\-—.,\s]\s*/i;

/** Caption text without a photo number typed into it; numbers are added when the report is made. */
export function stripNumber(text: string): string {
  const t = (text || '').trim();
  if (!t || PREFIX_ONLY.test(t)) return '';
  return t.replace(PREFIX, '').trim();
}

export function photoCaption(layout: PhotoLayout, number: number, text = ''): string {
  const body = stripNumber(text);
  const head = `${layout.caption_keyword} ${number}`;
  return body ? `${head} — ${body}` : head;
}

/** Photos split into report pages; a heading on the first page takes one row. */
export function photoPages<T>(photos: T[], layout: PhotoLayout): T[][] {
  if (!photos.length) return [];
  const first = layout.show_heading ? layout.per_page - 2 : layout.per_page;
  const pages: T[][] = [photos.slice(0, first)];
  for (let i = first; i < photos.length; i += layout.per_page) pages.push(photos.slice(i, i + layout.per_page));
  return pages.filter((p) => p.length);
}

/** Where a photo's image is served; the address changes when the photo is turned. */
export function photoUrl(reportId: string | undefined, assetId: string, asset: any): string {
  if (asset?.url) return asset.url;
  return reportId ? `/api/reports/${reportId}/assets/${assetId}/image` : '';
}
